ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ActiveContractsServiceIT.scala

```diff
@@ -394,6 +400,340 @@ class ActiveContractsServiceIT extends LedgerTestSuite {
+          case a :: Nil => f(a, a, a)
+          case invalid =>
+            throw new Exception(s"Invalid partySet, length must be 1 or 2 or 3 but it was $invalid")
+        }
+      val templateFactories: Vector[Set[Party] => Template[Template[Any]]] = Vector(
+        withThreeParties(TriAgreement(_, _, _)),
+        withThreeParties(TriProposal(_, _, _)),
+        parties => WithObservers(parties.head, parties.toList),
+      )
+      def createContractFor(template: Int, partyCombination: Int) =
+        ledger.create(
+          partyCombinations(partyCombination).map(parties).toList,
+          partyCombinations(partyCombination).map(parties).toList,
+          templateFactories(template)(partyCombinations(partyCombination).map(parties)),
+        )
+
+      val createFs = for {
+        partyCombinationIndex <- partyCombinations.indices
+        templateIndex <- templateFactories.indices
+        _ <- 1 to 3
+      } yield createContractFor(templateIndex, partyCombinationIndex)
+        .map((partyCombinationIndex, templateIndex) -> _)
+
+      Future
+        .sequence(createFs)
+        .map(_.groupBy(_._1).map { case ((partyCombinationIndex, templateIndex), contractId) =>
+          (
+            FilterCoord(templateIndex, partyCombinations(partyCombinationIndex)),
+            contractId.view.map(_._2.toString).toSet,
+          )
+        })
+    }
+
+    def testForFixtures(
+        fixtures: Vector[(Map[Int, Set[Int]], Set[FilterCoord])],
+        allContracts: Map[FilterCoord, Set[String]],
+    ) = {
+      def activeContractIdsFor(filters: Map[Int, Set[Int]]): Future[Vector[String]] =
+        ledger
+          .activeContracts(
+            new GetActiveContractsRequest(
+              ledgerId = ledger.ledgerId,
+              filter = Some(
+                new TransactionFilter(
+                  filters.map {
+                    case (party, templates) if templates.isEmpty =>
+                      (parties(party).toString, new Filters(None))
+                    case (party, templates) =>
+                      (
+                        parties(party).toString,
+                        new Filters(Some(new InclusiveFilters(templates.toSeq.map(templateIds)))),
+                      )
+                  }
+                )
+              ),
+              verbose = true,
+            )
+          )
+          .map(_._2.map(_.contractId))
+
+      def testForFixture(actual: Vector[String], expected: Set[FilterCoord], hint: String): Unit = {
+        val actualSet = actual.toSet
+        assert(
+          expected.forall(allContracts.contains),
+          s"$hint expected FilterCoord(s) which do not exist(s): ${expected.filterNot(allContracts.contains)}",
+        )
+        assert(
+          actualSet.size == actual.size,
+          s"$hint ACS returned redundant entries ${actual.groupBy(identity).toList.filter(_._2.size > 1).map(_._1).mkString("\n")}",
+        )
+        val errors = allContracts.toList.flatMap {
```

Comment:
That is exactly what I started with. Then I had test failures with the ACS implementation, which materialized itself in a list of contracts missing in the result. And then I experienced the pain to do manual mapping back from each missing contract to templateId/parties combination to have an understanding what is going on (in test cases there are at least some 30 contracts expected...say 5 is missing and 5 is extra...why those?). Then I realized what I was doing manually it is easy to do in code - this is it.
Now that it works it might make sense to simplify this back, but as soon this would start to fail we would face the same pain, so that's why I thought it is a good idea to keep this around.
With all that said I have no particularly strong opinion, although I would keep around this because I think debugging raw error would be much harder to follow than this code, and one have to do that every time when errors are there.
Happy to simplify back thought. What do you think @meiersi-da ?

Reply:
I can do that, no problem